In [1]:
import pandas as pd

df = pd.read_csv("Advertising.csv")
df.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [2]:
df = df.iloc[:, 1:len(df)]
df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [3]:
X = df.drop("sales", axis = 1)
y = df[["sales"]]
X.head()

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [4]:
y.head()

sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

In [6]:
# Statsmodels ile model kurmak

import statsmodels.api as sm

In [7]:
lm = sm.OLS(y, X)

In [8]:
model = lm.fit()

In [9]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  sales   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              3566.
Date:                Thu, 16 Jan 2025   Prob (F-statistic):                   2.43e-171
Time:                        21:50:16   Log-Likelihood:                         -423.54
No. Observations:                 200   AIC:                                      853.1
Df Residuals:                     197   BIC:                                      863.0
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0538      0.001     40.507      0.000       0.051       0.056
radio          0.2222      0.009     23.595      0.000       0.204       0.241
newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                         12.6
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### TABLO HAKKINDA BİLİNMESİ GEREKENLER

* R-square r2 degeri: bagimsiz degiskenlerin bagimli degiskenlerdeki degisikligi aciklama yuzdesi (satislardaki degisikligi aciklayan guzel degiskenler var)

* Adj R-square: ne kadar degisken eklenirse o kadar siser, duzeltilmis r2; daha saglikli bir aciklanabilirlik verir. %98 dogruluk var demek!

* prob(F-statistic): bu istatistik bize modelin anlamli mi degil mi onunla alakali bilgi verir 0.05 den dusukse anlamli

---
* coef: B1 B2 B3 ifadelerin katsayilarini ifade eder.

* std err: katsayıların standart hata

* P > |t|: katsayilar anlamli mi onu aciklar, 0.05 den kucukse anlamli oldugu kabul edilir

---

* TV deki 1 birim yatirim yapilmasi demek satislarda 0.05 artis olabilir demek
* Radiodaki 1 birim yatirim satislarda 0.2222 artis olabilir demek

In [11]:
## sicikit learn ile model kurmak

from sklearn.linear_model import LinearRegression
lm = LinearRegression()

In [12]:
model = lm.fit(X,y)

In [13]:
model.intercept_

array([2.93888937])

In [14]:
model.coef_
# coef parametreleri degisti. nedeni ise arka tarafta parametre tahmin yontemleri yuzunden!

array([[ 0.04576465,  0.18853002, -0.00103749]])

### Tahmin

#### Sales = 2.94 + TV * 0.04 + radio * 0.19 - newspaper * 0.001

30 birim TV, 10 birim radio, 40 birim gazete

In [16]:
2.94 + 30 * 0.04 + 10 * 0.19 - 40 * 0.001

5.999999999999999

In [17]:
yeni_veri = [[30],[10],[40]]

In [18]:
yeni_veri = pd.DataFrame(yeni_veri).T

In [19]:
yeni_veri

0   1   2
0  30  10  40

In [20]:
model.predict(yeni_veri) # biz ondalik degerleri attigimizdan biraz farkli sayi cikti

/home/berker/jupyter/jupyter_env/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[6.15562918]])

In [23]:
from sklearn.metrics import mean_squared_error

In [24]:
y.head()

sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

In [25]:
model.predict(X)[0:10]

array([[20.52397441],
       [12.33785482],
       [12.30767078],
       [17.59782951],
       [13.18867186],
       [12.47834763],
       [11.72975995],
       [12.12295317],
       [ 3.72734086],
       [12.55084872]])

In [27]:
MSE = mean_squared_error(y, model.predict(X))
MSE

2.784126314510936

In [28]:
import numpy as np

RMSE = np.sqrt(MSE)
RMSE

1.6685701407225697

### Model Tuning(Model Dogrulama)

In [29]:
X.head()

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [30]:
y.head()

sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

In [31]:
# sinama seti
from sklearn.model_selection import train_test_split

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [33]:
X_train.head()

TV  radio  newspaper
16    67.8   36.6      114.0
51   100.4    9.6        3.6
97   184.9   21.0       22.0
164  117.2   14.7        5.4
71   109.8   14.3       31.7

In [34]:
y_train.head()

sales
16    12.5
51    10.7
97    15.5
164   11.9
71    12.4

In [36]:
X_test.head()

TV  radio  newspaper
135  48.3   47.0        8.5
127  80.2    0.0        9.2
191  75.5   10.8        6.0
66   31.5   24.6        2.2
119  19.4   16.0       22.3

In [35]:
y_test.head()

sales
135   11.6
127    8.8
191    9.9
66     9.5
119    6.6

In [42]:
lm = LinearRegression()
model = lm.fit(X_train, y_train)

In [43]:
# Egitim hatasi
np.sqrt(mean_squared_error(y_train, model.predict(X_train)))

1.644727765644337

In [44]:
# Test hatasi
np.sqrt(mean_squared_error(y_test, model.predict(X_test)))

1.78159966153345

In [45]:
# K-Katli Cross Validation
# bunun yapilmasinin sebebi, train setine yonelik elde edecegimiz hatayi daha dogru degerlendirmek icin kullaniriz

# bu islem farkli varyasyonlara gore hata hesabı yapmak icin yapilir, dogrulanmis bir hata verir

In [49]:
from sklearn.model_selection import cross_val_score
cross_val_score(model, X_train, y_train, cv = 10, scoring = "neg_mean_squared_error")
# train seti uzerinden 10 farkli hata hesaplandi

array([-5.57303426, -2.86235681, -2.06504684, -1.09186983, -1.66159795,
       -2.50694042, -2.92821679, -2.01207197, -7.2250041 , -1.66156243])

In [50]:
# cv mse
np.mean(-cross_val_score(model, X_train, y_train, cv = 10, scoring = "neg_mean_squared_error"))
# ortalama

2.9587701386797574

In [52]:
# cv rmse
np.sqrt(np.mean(-cross_val_score(model, X_train, y_train, cv = 10, scoring = "neg_mean_squared_error")))

1.7201075950880973

In [53]:
# cv rmse
np.sqrt(np.mean(-cross_val_score(model, X, y, cv = 10, scoring = "neg_mean_squared_error")))
# veri seti kumesine gore cv rmse

1.749276312684338